src/transformers/models/gpt2/modeling_tf_gpt2.py

```diff
@@ -813,32 +813,29 @@ def get_output_embeddings(self):
     def set_output_embeddings(self, value):
         self.set_input_embeddings(value)
 
-    def prepare_inputs_for_generation(self, inputs, past=None, use_cache=None, use_xla=False, **kwargs):
-        # TODO: (Joao) after the TF generator is complete, update GPT2 TF generation to match PT's. NB -- some GPT2
-        # tests will need to be fixed after the change
-
+    def prepare_inputs_for_generation(self, inputs, past=None, use_cache=None, **kwargs):
+        token_type_ids = kwargs.get("token_type_ids", None)
         # only last token for inputs_ids if past is defined in kwargs
         if past:
             inputs = tf.expand_dims(inputs[:, -1], -1)
+            if token_type_ids is not None:
+                token_type_ids = tf.expand_dims(token_type_ids[:, -1], -1)
+
+        position_ids = kwargs.get("position_ids", None)
+        attention_mask = kwargs.get("attention_mask", None)
 
-        # TODO(pvp, Joao) - this `if use_xla` statement can be removed, but is left
-        # for a future PR to not change too many things for now.
-        # All statements in this if case apply for both xla and non-xla (as they already do in PyTorch)
-        position_ids = None
-        attention_mask = None
-        if use_xla:
-            attention_mask = kwargs.get("attention_mask", None)
-            if past is not None and attention_mask is not None:
-                position_ids = tf.reduce_sum(attention_mask, axis=1, keepdims=True) - 1
-            elif attention_mask is not None:
-                position_ids = tf.math.cumsum(attention_mask, axis=1, exclusive=True)
+        if attention_mask is not None and position_ids is None:
+            position_ids = tf.math.cumsum(attention_mask, axis=-1, exclusive=True)
+            if past:
+                position_ids = tf.expand_dims(position_ids[:, -1], -1)
```

Comment:
@gante 

I think this block should be 
```
        if attention_mask is not None and position_ids is None:
            position_ids = tf.math.cumsum(attention_mask, axis=-1, exclusive=True)
        if past and position_ids is not None:
            position_ids = tf.expand_dims(position_ids[:, -1], -1)
```
This is not a real issue, as I don't think `prepare_inputs_for_generation` is ever called with `position_ids`.

However, to make the logic more consistent, it's better to have processing similiar to `inputs` and `token_type_ids` (i.e. explicitly cut the sequence)

Don't feel urgent for this though.

Reply:
Sorry, after checking `GPT2LMHeadModel`, I believe it should be 
```
        if attention_mask is not None and position_ids is None:
            # create position_ids on the fly for batch generation
            ...
        else:
            position_ids = None
```
i.e. we should set `position_ids = None` in an `else` case.
The reason is that `position_ids` is never get updated in `update_model_kwargs_for_generation`.
(Although I don't really think `position_ids` will ever be provided to `model_kwargs`  for `generate`)